src/main/scala/strawman/collection/mutable/ArrayDeque.scala

```diff
@@ -0,0 +1,349 @@
+  set(array, start, end)
+
+  def this(initialSize: Int = ArrayDeque.DefaultInitialSize) = this(ArrayDeque.alloc(initialSize), 0, 0)
+
+  override def apply(idx: Int) = {
+    ArrayDeque.checkIndex(idx, this)
+    get(idx).asInstanceOf[A]
+  }
+
+  override def update(idx: Int, elem: A) = {
+    ArrayDeque.checkIndex(idx, this)
+    set(idx, elem.asInstanceOf[AnyRef])
+  }
+
+  override def +=(elem: A) = {
+    sizeHint(size)
+    appendAssumingCapacity(elem)
+  }
+
+  override def +=:(elem: A) = {
+    sizeHint(size)
+    prependAssumingCapacity(elem)
+  }
+
+  @inline private[ArrayDeque] def appendAssumingCapacity(elem: A): this.type = {
+    array(end) = elem.asInstanceOf[AnyRef]
+    end = end_-(-1)
+    this
+  }
+
+  @inline private[ArrayDeque] def prependAssumingCapacity(elem: A): this.type = {
+    start = start_+(-1)
+    array(start) = elem.asInstanceOf[AnyRef]
+    this
+  }
+
+  override def ++=:(elems: TraversableOnce[A]): this.type = {
+    // Note this is faster than super.++=: because this does not need to convert TraversableOnce to a Traversable
+    if (elems.isEmpty) return this
+    elems.sizeHintIfCheap match {
+      case srcLength if srcLength >= 0 && 2*(srcLength + this.length) >= mask =>
+        val finalLength = srcLength + this.length
+        val array2 = ArrayDeque.alloc(finalLength)
+        elems.copyToArray(array2.asInstanceOf[Array[A]])
+        copySliceToArray(srcStart = 0, dest = array2, destStart = srcLength, maxItems = size)
+        set(array = array2, start = 0, end = finalLength)
+      case _ =>
+        val copy = clone()
+        end = start
+        this ++= elems ++= copy
+    }
+    this
+  }
+
+  override def insertAll(idx: Int, elems: scala.collection.Traversable[A]): Unit = {
+    ArrayDeque.checkIndex(idx, this)
+    if (elems.isEmpty) return
+    elems.sizeHintIfCheap match {
+      case srcLength if srcLength >= 0 =>
+        val finalLength = srcLength + this.length
+        // Either we resize right away or move prefix right or suffix left
+        if (2*finalLength >= mask) {
+          val array2 = ArrayDeque.alloc(finalLength)
+          copySliceToArray(srcStart = 0, dest = array2, destStart = 0, maxItems = idx)
+          elems.copyToArray(array2.asInstanceOf[Array[A]], idx)
+          copySliceToArray(srcStart = idx, dest = array2, destStart = idx + srcLength, maxItems = size)
+          set(array = array2, start = 0, end = finalLength)
+        } else {
+          val suffix = drop(idx)
+          end = start_+(idx)
+          elems.foreach(appendAssumingCapacity)
+          suffix.foreach(appendAssumingCapacity)
+        }
+      case _ => //expensive to compute size
+        val suffix = drop(idx)
+        end = start_+(idx)
+        this ++= elems ++= suffix
+    }
+  }
+
+  override def remove(idx: Int, count: Int): Unit = {
+    ArrayDeque.checkIndex(idx, this)
+    if (count <= 0) return
```

Comment:
Well, I don't know, deleting nothing is a no-op so it probably shouldn't throw an exception?